### PR TITLE
added manual compute metrics, requires numpy still

### PR DESCRIPTION
**Pull request by ericlee0803**
This is an initial stab at computing the necessary classification / regression metrics when sklearn isn't installed. 

Will likely require revising so just issuing this PR right now to get feedback.

### CODE REVIEW

File: sigopt/xgboost/compute_metrics.py [hunk 1]
```diff
@@ -1,19 +1,76 @@
...
+      'support': support
+    }
+    classification_report['weighted avg']['precision'] += (support / len(y_pred)) * precision
+    classification_report['weighted avg']['recall'] += (support / len(y_pred)) * recall
+    classification_report['weighted avg']['f1-score'] += (support / len(y_pred)) * f1
+  return classification_report
+
+
+def compute_accuracy(y_true, y_pred):
+  accuracy = numpy.count_nonzero(y_true == y_pred) / len(y_true)
```
Comment (harvey-intel) [inline sigopt/xgboost/compute_metrics.py:49]: `numpy.sum` also works right?

File: sigopt/xgboost/compute_metrics.py [hunk 2]
```diff
@@ -29,8 +86,14 @@ def compute_regression_metrics(run, bst, D_matrix_pair):
...
+  if HAS_SKLEARN:
+    regression_metrics = {
+      f"{D_name}-mean absolute error": mean_absolute_error(y_test, preds),
+      f"{D_name}-mean squared error": mean_squared_error(y_test, preds)
+    }
+  else:
+    regression_metrics = {
+      f"{D_name}-mean absolute error": compute_mae(y_test, preds),
+      f"{D_name}-mean squared error": compute_mse(y_test, preds)
+    }
```
Comment (harvey-intel) [inline sigopt/xgboost/compute_metrics.py:98]: Do the `if` block inside the `compute_mae`, same for the other metrics.
```
def compute_mae(y_true, y_pred, has_sklearn=True):
  if has_sklearn:
    return mean_absolute_error(....)

  return numpy.mean(abs(...))
 ```

File: sigopt/xgboost/compute_metrics.py (hunk 1 above)
Comment (harvey-intel) [reply sigopt/xgboost/compute_metrics.py:49]: looks like `count_nonzero` is more efficient according to stackoverflow

File: sigopt/xgboost/compute_metrics.py [hunk 3]
```diff
@@ -1,19 +1,76 @@
...
+  HAS_SKLEARN = True
+except ImportError as e:
+  HAS_SKLEARN = False
+
+
+def compute_positives_and_negatives(y_true, y_pred, class_label):
+  tp = numpy.count_nonzero(numpy.logical_and(y_true == class_label, y_pred == class_label))
+  tn = numpy.count_nonzero(numpy.logical_and(y_true != class_label, y_pred != class_label))
+  fp = numpy.count_nonzero(numpy.logical_and(y_true != class_label, y_pred == class_label))
+  fn = numpy.count_nonzero(numpy.logical_and(y_true == class_label, y_pred != class_label))
```
Comment (harvey-intel) [inline sigopt/xgboost/compute_metrics.py:18]: you can slightly improve this by pre-compute the `y_true == class_label` and `y_pred == class_label` and just reuse them.

File: sigopt/xgboost/compute_metrics.py [hunk 4]
```diff
@@ -1,19 +1,76 @@
...
 def compute_classification_metrics(run, bst, D_matrix_pair):
   D_matrix, D_name = D_matrix_pair
   preds = bst.predict(D_matrix)
   preds = numpy.round(preds)
   y_test = D_matrix.get_label()
-  accuracy = accuracy_score(y_test, preds)
-  rep = classification_report(y_test, preds, output_dict=True, zero_division=0)
+  if HAS_SKLEARN:
+    accuracy = accuracy_score(y_test, preds)
+    rep = classification_report(y_test, preds, output_dict=True, zero_division=0)
```
Comment (harvey-intel) [inline sigopt/xgboost/compute_metrics.py:70]: how do we deal with the zero divide when not using sklearn?

File: sigopt/xgboost/compute_metrics.py [hunk 5]
```diff
@@ -1,19 +1,76 @@
...
+
+
+def compute_mae(y_true, y_pred):
+  d = y_true - y_pred
+  return numpy.mean(abs(d))
+
+
+def compute_mse(y_true, y_pred):
+  d = y_true - y_pred
+  return numpy.mean(d ** 2)
```
Comment (harvey-intel) [inline sigopt/xgboost/compute_metrics.py:60]: We need unit tests for these functions.

**Review by harvey-intel — CHANGES_REQUESTED** [review]
looks mostly right to me, but we need unit tests.

File: sigopt/xgboost/compute_metrics.py [hunk 6]
```diff
@@ -1,19 +1,76 @@
...
+try:
+  from sklearn.metrics import (
+    accuracy_score,
+    classification_report,
+    mean_absolute_error,
+    mean_squared_error
+  )
+  HAS_SKLEARN = True
+except ImportError as e:
+  HAS_SKLEARN = False
```
Comment (yuhong-sigopt) [inline sigopt/xgboost/compute_metrics.py:11]: It is better to define all the missing metric functions here, so you do need need to change the code where metric functions are used.

File: sigopt/xgboost/compute_metrics.py (hunk 4 above)
Comment (ericlee0803) [reply sigopt/xgboost/compute_metrics.py:70]: forgot about this, thanks!

File: sigopt/xgboost/compute_metrics.py (hunk 6 above)
Comment (ericlee0803) [reply sigopt/xgboost/compute_metrics.py:11]: How exactly do you suggest doing this?

Comment (yuhong-sigopt) [reply sigopt/xgboost/compute_metrics.py:11]: something like:
`
except ImportError as e:
   def mean_absolute_error(y_true, y_pred):
     d = y_true - y_pred
     return numpy.mean(abs(d))

    
`

Comment (ericlee0803) [reply sigopt/xgboost/compute_metrics.py:11]: I don't like this because it overloads the function and prevents easy unit testing to compare against the sklearn version.

Comment (yuhong-sigopt) [reply sigopt/xgboost/compute_metrics.py:11]: You can write the metric functions in another file for test, change the code about as
```
except ImportError as e:
   import utils.mean_absolute_error as mean_absolute_error
```

Comment (ericlee0803) [reply sigopt/xgboost/compute_metrics.py:11]: I see. Let me revise everything else, write unit tests, and let's get back to this after!

**Comment by ericlee0803** [comment]
Now that I'm thinking about it... is there any particular reason to use sklearn? Seems easier to just use all the internal code, makes the code cleaner

**Comment by harvey-intel** [comment]
> Now that I'm thinking about it... is there any particular reason to use sklearn? Seems easier to just use all the internal code, makes the code cleaner

not particular, I think. We use the sklearn dataset in integration test, but we can get around that as well by generating our own datasets.

File: integration_test/test_compute_metrics.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    # Check classification metrics
+    assert np.abs(report_compute['weighted avg']['precision'] - report_sklearn['weighted avg']['precision']) < 1e-8
+    assert np.abs(report_compute['weighted avg']['recall'] - report_sklearn['weighted avg']['recall']) < 1e-8
+    assert np.abs(report_compute['weighted avg']['f1-score'] - report_sklearn['weighted avg']['f1-score']) < 1e-8
+    assert np.abs(_compute_accuracy(Y_test, preds) - accuracy_score(Y_test, preds)) < 1e-8
+
+  def test_regression_metrics(self):
+    # Check regression metrics
+    assert np.abs(mean_absolute_error(Y_test, preds) - compute_mae(Y_test, preds)) < 1e-8
+    assert np.abs(mean_squared_error(Y_test, preds) - compute_mse(Y_test, preds)) < 1e-8
```
Comment (harvey-intel) [inline integration_test/test_compute_metrics.py:51]: I do want to consider more test coverage here. Testing the metric computation does not require us to actually train an xgb model. We can create whatever Y_pred array we want. So it's best to manually create a few Y_test/Y_pred arrays to test out a few different classification scenarios.
1. binary classification / multiclass classification (may also want to consider class imbalance)
2. Y_pred only predicts 1 class
3. Y_pred completely misclassifies one class (related to class imbalance)

Regression metrics are easier to test, you can just create a Y_test numpy array and compute the mae/mse against a few different known arrays, e.g., itself, all-zero, some constant offsite, some other known arrays.

Comment (ericlee0803) [reply integration_test/test_compute_metrics.py:51]: agreed!

File: integration_test/test_compute_metrics.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    y_pred = np.zeros(n)
+    assert_classification_metrics(y_true, y_pred)
+
+  def test_zero_divide_binary(self):
+    n = 50
+    y_true = np.random.randint(0, 2, n)
+    y_pred = np.zeros(n)
+    assert_classification_metrics(y_true, y_pred)
+
+  def test_regression_metrics(self):
```
Comment (harvey-intel) [inline integration_test/test_compute_metrics.py:84]: This is ok. I was thinking along the line of deterministic Y_true/Y_pred arrays to test these functions. Because you can just create whatever ground truth you want. But this is ok too if we want to use sklearn as the groundtruth for unit test.

File: integration_test/test_compute_metrics.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
+from sigopt.xgboost.compute_metrics import (
```
Comment (harvey-intel) [inline integration_test/test_compute_metrics.py:1]: I think this file should live under `test/` direction, since it's *not* an integration test.

File: integration_test/test_compute_metrics.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    params = {
+      'eta': 0.3,
+      'num_class': num_class,
+      'objective': 'multi:softmax'
+    }
+    bst = xgb.train(params, D_train)
+    preds = bst.predict(D_test)
+    xgb_preds = np.round(preds)
+    xgb_true = Y_test
+    assert_classification_metrics(xgb_true, xgb_preds)
```
Comment (harvey-intel) [inline integration_test/test_compute_metrics.py:56]: I think this portion can live in integration_test.

**Review by harvey-intel — COMMENTED** [review]
A few more nits.

File: integration_test/test_compute_metrics.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+  compute_accuracy
+)
+from sklearn.metrics import (
+  accuracy_score,
+  classification_report,
+  mean_absolute_error,
+  mean_squared_error
+)
+from sklearn import datasets
+from sklearn.model_selection import train_test_split
```
[the remark this answers is not in the document]
Comment (dwanderson-intel) [reply integration_test/test_compute_metrics.py:14]: Correct - if it's just for developers / testing, then it goes into `requirements-dev.txt`.

Looks like that file (dev.txt) gets optionally included into setup.py (`extra requires`) but I don't quite know how that works - presumably it's fine though to just have it there.

File: sigopt/xgboost/compute_metrics.py [hunk 12]
```diff
@@ -1,19 +1,67 @@
...
+    support = numpy.count_nonzero(y_true == class_label)
+    classification_report[str(class_label)] = {
+      'precision': precision,
+      'recall': recall,
+      'f1-score': f1,
+      'support': support
+    }
+    classification_report['weighted avg']['precision'] += (support / len(y_pred)) * precision
+    classification_report['weighted avg']['recall'] += (support / len(y_pred)) * recall
+    classification_report['weighted avg']['f1-score'] += (support / len(y_pred)) * f1
```
[the remark this answers is not in the document]
Comment (ericlee0803) [reply sigopt/xgboost/compute_metrics.py:43]: That's true, but it does seem like an reasonable way of storing these metrics... inclined to keep unless you have another suggestion

File: integration_test/test_compute_metrics.py (hunk 10 above)
Comment (ericlee0803) [reply integration_test/test_compute_metrics.py:56]: Actually can I just delete this test? Doesn't seem necessary tbh

Comment (ericlee0803) [reply integration_test/test_compute_metrics.py:56]: That way we can just have all the compute_metrics tests in one place

File: sigopt/xgboost/compute_metrics.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (harvey-intel) [reply sigopt/xgboost/compute_metrics.py:43]: Since we don't report/log the per-class metrics here, storing them seems unnecessary. However, I do see a case where maybe we want to return per-class metric for binary-classification problems, then this is probably fine as is.

File: integration_test/test_compute_metrics.py (hunk 10 above)
Comment (harvey-intel) [reply integration_test/test_compute_metrics.py:56]: Yea, you can. You will still need to add some tests to the integration_test so that we know we can verify `run.get_metric` are returning the appropriate values.

File: integration_test/test_compute_metrics.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (ericlee0803) [reply integration_test/test_compute_metrics.py:14]: I've updated the unit tests to check for sklearn (see tests/test_compute_metrics.py), if I do this then we don't need to add Scikit-Learn to the requirements-dev.txt at all. But is this necessary?

[the remark this answers is not in the document]
Comment (dwanderson-intel) [reply integration_test/test_compute_metrics.py:14]: not sure i follow the `test/test_compute_metrics.py` and "check for sklearn" - is that a different PR or something?

[the remark this answers is not in the document]
Comment (ericlee0803) [reply integration_test/test_compute_metrics.py:14]: never mind! I was not thinking straight

**Comment by ericlee0803** [comment]
This ended up being a bigger PR than expected, due to all the tests. Sorry about that! Please let me know if there's something I can do to compact them down.

File: integration_test/test_compute_correct_metrics.py [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+      'eta': 0.3,
+      'num_class': num_class,
+      'objective': 'multi:softmax'
+    }
+    xgb_params = {
+      'params': params,
+      'dtrain': D_train,
+      'evals': [(D_test, 'test0')],
+    }
+    ctx = sigopt.xgboost.run(**xgb_params)
```
Comment (yuhong-sigopt) [inline integration_test/test_compute_correct_metrics.py:34]: If just test compute metrics, it is no need to train a model and predict.

Comment (ericlee0803) [reply integration_test/test_compute_correct_metrics.py:34]: That is a fair point. I do think this test is somewhat unnecessary, but it does test the metric logging correctness. Whereas test/test_compute_metrics.py tests a few different edge cases when computing these metrics. The tests could possibly merged, thoughts @harvey-intel ?

Comment (harvey-intel) [reply integration_test/test_compute_correct_metrics.py:34]: Yea, it's a good idea to consolidate these into the `TestXGBoost` in `integration_test/test_xgboost.py`.

Comment (ericlee0803) [reply integration_test/test_compute_correct_metrics.py:34]: Okay, so then I will remove test/test_compute_metrics.py and move tests in here, and also remove the sklearn dependency in requirements-dev.txt, is that okay?

Comment (harvey-intel) [reply integration_test/test_compute_correct_metrics.py:34]: Yea, I would consolidate everything here and `test_log_metric.py` under `TestXGBoost` in `integration_test/test_xgboost.py`.

File: setup.py [hunk 14]
```diff
@@ -28,7 +28,7 @@
 with open(os.path.join(here, 'requirements-dev.txt')) as requirements_dev_fp:
   dev_install_requires = requirements_dev_fp.read().split('\n')
 
-xgboost_install_requires = ['xgboost>=1.3.1', 'scikit-learn>=0.23.2', 'numpy>=1.15.0']
+xgboost_install_requires = ['xgboost>=1.3.1', 'numpy>=1.15.0']
```
Comment (harvey-intel) [inline setup.py:31]: scikit-learn needs to be in the requirements-dev.txt.

File: integration_test/test_compute_metrics.py [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+    y_true = np.random.randint(0, 2, n)
+    y_pred = np.zeros(n)
+    assert_classification_metrics(y_true, y_pred)
+
+  def test_regression_metrics(self):
+    # Check regression metrics
+    n = 100
+    y_true = np.random.randn(n)
+    y_pred = np.random.randn(n)
+    assert_regression_metrics(y_true, y_pred)
```
[the remark this answers is not in the document]
Comment (harvey-intel) [reply integration_test/test_compute_metrics.py:130]: this is still here lol.

File: integration_test/test_compute_metrics.py [hunk 16]
```diff
@@ -0,0 +1,130 @@
+import sigopt.xgboost
```
Comment (harvey-intel) [inline integration_test/test_compute_metrics.py:1]: this file needs to be under `test/` directory.

File: integration_test/test_compute_metrics.py [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+    preds = bst.predict(D_test)
+    xgb_preds = np.round(preds)
+    xgb_true = Y_test
+    mae = mean_absolute_error(xgb_true, xgb_preds)
+    mse = mean_squared_error(xgb_true, xgb_preds)
+
+    # compare metrics
+    run = sigopt.get_run(ctx.run.id)
+    assert np.abs(run.values['test0-mean absolute error'].value - mae) < (1e-8 / mae)
+    assert np.abs(run.values['test0-mean squared error'].value - mse) < (1e-8 / mae)
```
Comment (harvey-intel) [inline integration_test/test_compute_metrics.py:95]: move all these into `integration_test`

Comment (ericlee0803) [reply integration_test/test_compute_metrics.py:95]: Okay I think there was a misunderstanding. So right now we have three sets of tests corresponding to three classes: 

- TestComputeXGBMetrics (check if xgb.run records correct metrics)
- TestComputeSyntheticMetrics (checks a variety of metric computations using synthetic stuff)
- TestLogMetrics (just checks logging)

Which of these ought to be in integration_test/ and which of these ought to be in test/ ?

File: sigopt/xgboost/compute_metrics.py [hunk 18]
```diff
@@ -1,19 +1,66 @@
...
```
Comment (yuhong-sigopt) [inline sigopt/xgboost/compute_metrics.py:43]: I think `weighted avg` is not necessary.

**Review by yuhong-sigopt — APPROVED** [review]

File: requirements-dev.txt [hunk 19]
```diff
@@ -12,3 +12,4 @@ pylint==2.9.6
 pytest==6.2.1
 setuptools>=47.3.1
 twine>=3.2.0,<4.0.0
+scikit-learn>=0.23.
```
Comment (dwanderson-intel) [inline requirements-dev.txt:15]: typo? presumably needs a patch version

Comment (ericlee0803) [reply requirements-dev.txt]: Thank you so m much for catching this!! Fixed
